# **Extract, Transform and Load (ETL)**

### Objectives

The purpose for this notebook is to extract, transform and load a raw dataset which will be saved into a clean data frame for Exploratory Data Analysis process. 

1. **Extract** healthcare insurance dataset. Load the raw data into this notebook to begin summarising the DataFrame by applying Pandas methods. 
2. **Transform** check for missing or unique values and duplicated rows to ensure the data is clean prior to further analysis. Standardise text data to ensure consistency for grouping by converting text to lower-case and removing whitespace. Engineering DataFrames adding categorical columns to group together variables.
3. **Load** convert and save clean healthcare insurance DataFrame into a csv file to load in the data visualisation notebook. 

### Input

A CSV healthcare insurance dataset is stored in the Data folder. The dataset is required for the analysis and contains healthcare insurance information that will be used to identify patterns, relationships, and trends within the data.

> File: healthcare_insurance.csv

>File path: /Users/admin/Documents/Code Institute/vscode-projects/Healthcare-Analysis-Project/Healthcare-Analysis-Project/Data/healthcare_insurance.csv

### Import Libraries

Importing dependencies to extract, transform and load data frames for this project including Numpy and Pandas. 

In [1]:
# Importing libraries installed and added to requirements.txt
import numpy as np # Import numpy to perform numerical calculations
import pandas as pd # Import pandas to extract data

## **Extract DataFrame**

### **Load DataFrame**

Load CSV healthcare insurance dataset using Pandas. I will be referencing the cheatsheet you can access on the link below.

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

In [2]:
# Load healthcare_insurance.csv file path 
df = pd.read_csv('/Users/admin/Documents/Code Institute/vscode-projects/Healthcare-Analysis-Project/Healthcare-Analysis-Project/Data/healthcare_insurance.csv')
# Check first 5 rows in the dataframe
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


### **Inspect DataFrame**

The code cells below present a summary of the dataframe from different points of view. This intial inspection will help identify attributes which could reduce the quality of the DataFrame.

<br/>

##### **DataFrame Shape**

Run <code>df.shape</code> in code cell to check the size of the DataFrame.

In [3]:
df.shape # Check number of rows and columns 

(1338, 7)

The DataFrame contains 1338 rows, 7 columns. 

<br/>

##### **DataFrame Info**

Run <code>df.info()</code> in code cell to print a summary for the DataFrame. 

In [4]:
df.info() # Concise summary for the datset 

<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB


From this point of view we can summarise the DataFrame as having the folowing attributes: 

* An index ranging between 0 to 1337.

* There are no missing values in each column. 

* Two columns containing decimal numbers (float64), two columns containing whole numbers (int64) and three columns containing text (str). 

* The memory usage is 73.3 KB.

<br/>

##### **DataFrame Columns (Titles)**

Run <code>df.columns</code> to highlight any unsual or incoherent formatting which could be standardised by stripping whitespace and converting text strings into lowercase. 

In [5]:
df.columns # Check column names

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='str')

The output for <code>df.columns</code> has informed me I do not need to format column names. 

<br/>

##### **DataFrame Columns (Statistical Overview)**

Run <code>df.describe()</code> to print an overview for each column with descriptive and statistical details.

In [6]:
df.describe() # Descriptive and statistical overview for each column

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


The summary highlights numerical attributes contained in the DataFrame that could impact the realiability of the data in the EDA process. 

<br/>

##### **Unique Values**

Run <code>df.nununique()</code> to check the DataFrame for unique values.  

In [7]:
df.nunique() # Check for unique values in DataFrame

age           47
sex            2
bmi          548
children       6
smoker         2
region         4
charges     1337
dtype: int64

Based on the attributes printed I do not suspect there are any errors in the code. However, previously I became aware there are 1338 rows and I can see there is 1337/1338 unique charges meaning there could be a duplicated value worth inspecting closer.

<br/>

##### **Duplicated (Rows)**

Run <code> df.duplicated()</code>  to check the DataFrame for duplicated rows.

In [8]:
df.duplicated().value_counts() # Check for duplicated rows in columns 

False    1337
True        1
Name: count, dtype: int64

In [9]:
df[df.duplicated('charges', keep=False) == True] # Print rows sharing duplicated value

,age,sex,bmi,children,smoker,region,charges
195,19,male,30.59,0,no,northwest,1639.5631
581,19,male,30.59,0,no,northwest,1639.5631


In [10]:
df.iloc[195] # Print row 195 to double-check values 

age                19
sex              male
bmi             30.59
children            0
smoker             no
region      northwest
charges     1639.5631
Name: 195, dtype: object

In [11]:
df.iloc[581] # Print row 581 to double-check values 

age                19
sex              male
bmi             30.59
children            0
smoker             no
region      northwest
charges     1639.5631
Name: 581, dtype: object

<br/>

#### **Summary**

At first sight the DataFrame appeared to be clean. There was nothing outstanding in the initial inspection. Column names were tidy and thus did not require formatting. 

However, an unexpected duplicated row was found afterall. In the next stage of the ETL pipeline I will create a clean DataFrame. 

<br/>

## **Transform DataFrame**

In the previous section I encountered a duplicated row. To meet the business requirements for the Healthcare Insurance Analysis it is neccessary to enhance the instegrity of the data during the EDA process by creating a clean DataFrame instead of the raw data I downloaded from Kaggle. 

<br/>

### **Drop Duplicate Rows**

Remove duplicate row identified during the DataFrame summary. Create a copy of the original DataFrame and apply <code> df.drop_duplicates()</code> method. 

Check the number of rows in the DataFrame copy below the original DataFrame shape to verify the duplicate row has been dropped using <code>df.shape</code> and <code>df_copy.shape</code>.

In [12]:
# Create copy of original DataFrame and apply drop duplicates method 
df_copy = df.drop_duplicates()

print (df.shape) # Print shape for original DataFrame
print (df_copy.shape) # Print shape for DataFrame copy 


(1338, 7)
(1337, 7)


The results show evidence the <code>df.drop_duplicates</code> method has been applied to the  <code>df_copy</code>. 

<br/>

### **Feature Engineering**

Create features using the DataFrame copy to streamline and improve the EDA process. In this section I will create two DataFrame transforming categorical values into numerical insights to questions listed in the hypothesis.

* How does smoking status affect the minimum, average and maximum insurance charges?

* Can healthcare insurance charges be predicted by BMI?

These two questions contribute to predictive analysis but could be developed further to become a machine learning feature. 

<br/>

#### **Aggregate Categorical Variable**

Create a DataFrame to apply predictive analysis on smokers status. Add columns including minimum, average and maximum healthcare insurance charges. 

In [13]:
df_copy.head(4)

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061


Create DataFrame featuring minimum, average and maximum charges between smokers and non-smoker. 

In [18]:
# Creat DataFrame for smoking status charges
 
df_smoking_charges = df_copy.groupby('smoker')['charges'].agg(['min', 'mean', 'max'])
df_smoking_charges


,min,mean,max
smoker,,,
no,1121.8739,8440.660307,36910.60803
yes,12829.4551,32050.231832,63770.42801


Note: CoPilot fixed the code inside the code cell to create the DataFrame. 

<br/>